src/rules/selector-max-depth/index.js

```diff
@@ -0,0 +1,78 @@
+import resolvedNestedSelector from "postcss-resolve-nested-selector"
+
+import {
+  isStandardRule,
+  isStandardSelector,
+  report,
+  ruleMessages,
+  validateOptions,
+} from "../../utils"
+
+export const ruleName = "selector-max-depth"
+
+export const messages = ruleMessages(ruleName, {
+  expected: (selector, depth) => `Expected "${selector}" to have a depth equal to or less than ${depth}`,
+})
+
+export default function (max) {
+  return (root, result) => {
+    const innerSelectorReg = /(?::not\((.*?)\))/g
+    const validOptions = validateOptions(result, ruleName, {
+      actual: max,
+      possible: [function (max) {
+        return typeof max === "number" && max > 0
+      }],
+    })
+    if (!validOptions) { return }
+
+    // Compares depth of a selector with the maximum
+    function checkDepth(selector, rule) {
```

Comment:
> To be more about "we need to use standard parsers whenever possible... because there are significant gains in doing so".

Yep, I meant exactly that part.

Thanks everyone for talking this out.